Modules/timemodule.c

```diff
@@ -834,101 +824,127 @@ time_strftime(PyObject *module, PyObject *args)
-        }
-#if defined _MSC_VER && _MSC_VER >= 1400 && defined(__STDC_SECURE_LIB__)
-        errno = 0;
-#endif
-        _Py_BEGIN_SUPPRESS_IPH
-        buflen = format_time(outbuf, i, fmt, &buf);
-        _Py_END_SUPPRESS_IPH
-#if defined _MSC_VER && _MSC_VER >= 1400 && defined(__STDC_SECURE_LIB__)
-        /* VisualStudio .NET 2005 does this properly */
-        if (buflen == 0 && errno == EINVAL) {
-            PyErr_SetString(PyExc_ValueError, "Invalid format string");
-            PyMem_Free(outbuf);
-            break;
-        }
-#endif
-        if (buflen > 0 || i >= 256 * fmtlen) {
-            /* If the buffer is 256 times as long as the format,
-               it's probably not failing for lack of room!
-               More likely, the format yields an empty result,
-               e.g. an empty format, or %Z when the timezone
-               is unknown. */
+            /* I hate these functions that presume you know how big the output
+             * will be ahead of time...
+             */
+            while (1) {
+                outbuf = (time_char *)PyMem_Realloc(outbuf, bufsize*sizeof(time_char));
+                if (outbuf == NULL) {
+                    PyErr_NoMemory();
+                    goto error;
+                }
+    #if defined _MSC_VER && _MSC_VER >= 1400 && defined(__STDC_SECURE_LIB__)
+                errno = 0;
+    #endif
+                _Py_BEGIN_SUPPRESS_IPH
+                buflen = format_time(outbuf, bufsize, format, &buf);
+                _Py_END_SUPPRESS_IPH
+    #if defined _MSC_VER && _MSC_VER >= 1400 && defined(__STDC_SECURE_LIB__)
+                /* VisualStudio .NET 2005 does this properly */
+                if (buflen == 0 && errno == EINVAL) {
+                    PyErr_SetString(PyExc_ValueError, "Invalid format string");
+                    goto error;
+                }
+    #endif
+                if (buflen == 0 && bufsize < 256 * fmtlen) {
+                    bufsize += bufsize;
+                    continue;
+                }
+                /* If the buffer is 256 times as long as the format,
+                   it's probably not failing for lack of room!
+                   More likely, the format yields an empty result,
+                   e.g. an empty format, or %Z when the timezone
+                   is unknown. */
+                break;
+            }
+            PyObject *unicode;
 #ifdef HAVE_WCSFTIME
-            ret = PyUnicode_FromWideChar(outbuf, buflen);
+            unicode = PyUnicode_FromWideChar(outbuf, buflen);
 #else
-            ret = PyUnicode_DecodeLocaleAndSize(outbuf, buflen, "surrogateescape");
+            unicode = PyUnicode_DecodeLocaleAndSize(outbuf, buflen, "surrogateescape");
 #endif
-            PyMem_Free(outbuf);
-            break;
+            if (_PyUnicodeWriter_WriteStr(&writer, unicode) < 0) {
+                Py_DECREF(unicode);
+                goto error;
+            }
+            Py_DECREF(unicode);
+        }
+
+        j = i;
```

Comment:
You should declare the `j` variable here, and I suggest to rename it to `start`.